feat(lua): vim.version() returns Version object

mg979 opened the pull request:
- vim.version() returns a Version object. Makes it printable and removes the need of workarounds when passing it to other vim.version methods.

- ~~vim.version.range() converts a table to a Version object only when necessary.~~

-~~ vim.version.is() To easily check a version range against current version, or target version. Eg.:~~

      vim.version.is('>0.9.0')          --> true
      vim.version.is('>0.9.0', '0.8.0') --> false

- ~~cache the result of vim.fn.api_info().version~~

mg979 committed:
feat: vim.version() returns a Version object

- vim.version() returns a Version object.
  Makes it printable and removes the need of workarounds when passing it
  to other vim.version methods.

justinmk commented:
> * vim.version.is() To easily check a version range against current version, or target version. Eg.:

No, this is too much. We already have `vim.fn.has('nvim-1.2.3')`. Don't add more surface area.

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2734,6 +2734,24 @@ versions (1.2.3-rc1) are not matched. >
 
 <
 
+is({spec}, {version})                                       *vim.version.is()*
```
remove. not worth the extra surface area, docs, tests, etc.

justinmk commented on runtime/lua/vim/version.lua:
```diff
@@ -422,9 +437,24 @@ function M.parse(version, opts)
 end
 
+-- Cached vim.fn.api_info().version.
+local _api_v
```
no

mg979 commented:
> vim.version() returns a Version object. Makes it printable and removes the need of workarounds when passing it to other vim.version methods.
vim.version.range() converts a table to a Version object only when necessary.

Are these ok?

justinmk commented on runtime/lua/vim/version.lua:
```diff
@@ -423,8 +421,12 @@ function M.parse(version, opts)
 end
 
 setmetatable(M, {
+  --- Return a Version object, representing the current neovim version.
```
```suggestion
  --- Return the current Nvim version.
```

justinmk commented on runtime/lua/vim/version.lua:
```diff
@@ -423,8 +421,12 @@ function M.parse(version, opts)
-    return vim.fn.api_info().version
+    local version = vim.fn.api_info().version
+    -- Workaround: vim.fn.api_info().version reports "prerelease" as a boolean.
+    version.prerelease = version.prerelease or nil
```
If prerelease is `true` will that be a problem?

justinmk commented:
> vim.version() returns a Version object. Makes it printable

That would fix https://github.com/neovim/neovim/issues/23863 .
I see we have `Version:__tostring()` https://github.com/neovim/neovim/blob/ecdb6465e272119f80f3b68a6695e0f74b02ca49/runtime/lua/vim/version.lua#L123  but with this PR I still see:

```
:lua tostring(vim.version())
table: 0x010cad9980
```

justinmk committed:
feat: tostring(vim.version())

Problem:
tostring(vim.version()) returns "table: 0x…".

Solution:
Modify vim.version() to return a string prerelease instead of a boolean.

Fix #23863

justinmk commented on runtime/lua/vim/version.lua:
```diff
@@ -423,8 +421,12 @@ function M.parse(version, opts)
-    return vim.fn.api_info().version
+    local version = vim.fn.api_info().version
+    -- Workaround: vim.fn.api_info().version reports "prerelease" as a boolean.
+    version.prerelease = version.prerelease and 'dev' or nil
```
Hack: just return "dev" for now, so that at least the prerelease field is properly a string. Later maybe someone can find a way to get the actual prerelease string.
